node/router/src/heartbeat.rs

```diff
@@ -233,11 +237,19 @@ pub trait Heartbeat<N: Network>: Outbound<N> {
         if num_deficient > 0 {
             // Initialize an RNG.
             let rng = &mut OsRng;
+            let banned_ips = self.tcp().banned_peers().get_banned_ips();
 
             // Attempt to connect to more peers.
-            for peer_ip in self.router().candidate_peers().into_iter().choose_multiple(rng, num_deficient) {
+            for peer_ip in self
+                .router()
+                .candidate_peers()
```

Comment:
I wonder if it would make sense to exclude the banned IPs directly from `candidate_peers`